File: backend/app/schemas/media.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.media_asset import MediaType


class MediaMetadataBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    media_type: MediaType
    object_key: str = Field(min_length=1, max_length=1024)
    cdn_url: str | None = Field(default=None, min_length=1, max_length=2048)
    version: int = Field(default=1, ge=1)
    file_size: int = Field(ge=0)


class MediaCreate(MediaMetadataBase):
    id: uuid.UUID | None = None
    content_type: str = Field(min_length=1, max_length=255)


class MediaCreateMetadata(MediaMetadataBase):
    id: uuid.UUID | None = None


class MediaUploadUrlRequest(BaseModel):
    media_id: uuid.UUID | None = None
    version: int = Field(default=1, ge=1)
    filename: str = Field(min_length=1, max_length=255)
    content_type: str = Field(min_length=1, max_length=255)
    file_size: int = Field(gt=0)


class MediaUploadUrlResponse(BaseModel):
    media_id: uuid.UUID
    version: int
    upload_url: str
    object_key: str
    expires_in: int
    required_headers: dict[str, str]


class MediaPlaybackUrlResponse(BaseModel):
    media_id: uuid.UUID
    version: int
    playback_url: str
    expires_in: int


class MediaUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    media_type: MediaType | None = None
    object_key: str | None = Field(default=None, min_length=1, max_length=1024)
    cdn_url: str | None = Field(default=None, min_length=1, max_length=2048)
    version: int | None = Field(default=None, ge=1)
    file_size: int | None = Field(default=None, ge=0)


class MediaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    media_type: MediaType
    version: int
    file_size: int
    created_at: datetime
    updated_at: datetime


class MediaList(BaseModel):
    items: list[MediaRead]
    total: int
    limit: int
    offset: int

File: frontend/src/api/media.ts
import { apiRequest } from "./client";

export type MediaType = "image" | "video" | "audio" | "document" | "other";

export type MediaRead = {
  id: string;
  title: string;
  media_type: MediaType;
  version: number;
  file_size: number;
  created_at: string;
  updated_at: string;
};

export type MediaList = {
  items: MediaRead[];
  total: number;
  limit: number;
  offset: number;
};

export type MediaUploadUrlRequest = {
  media_id?: string;
  version?: number;
  filename: string;
  content_type: string;
  file_size: number;
};

export type MediaUploadUrlResponse = {
  media_id: string;
  version: number;
  upload_url: string;
  object_key: string;
  expires_in: number;
  required_headers: Record<string, string>;
};

export type MediaCreate = {
  id?: string;
  title: string;
  media_type: MediaType;
  object_key: string;
  cdn_url?: string | null;
  version: number;
  file_size: number;
  content_type: string;
};

export type MediaPlaybackUrlResponse = {
  media_id: string;
  version: number;
  playback_url: string;
  expires_in: number;
};

export function listMedia(params: { limit?: number; offset?: number } = {}) {
  const searchParams = new URLSearchParams();
  if (params.limit !== undefined) {
    searchParams.set("limit", String(params.limit));
  }
  if (params.offset !== undefined) {
    searchParams.set("offset", String(params.offset));
  }
  const query = searchParams.toString();
  return apiRequest<MediaList>(`/media${query ? `?${query}` : ""}`);
}

export function requestMediaUploadUrl(payload: MediaUploadUrlRequest) {
  return apiRequest<MediaUploadUrlResponse>("/media/upload-url", {
    method: "POST",
    body: payload
  });
}

export function createMedia(payload: MediaCreate) {
  return apiRequest<MediaRead>("/media", {
    method: "POST",
    body: payload
  });
}

export function requestMediaPlaybackUrl(mediaId: string) {
  return apiRequest<MediaPlaybackUrlResponse>(`/media/${mediaId}/playback-url`, {
    method: "POST"
  });
}
